### ISSUE TITLE
Compiler version constraint is not properly accounted with multiple compilers

### ISSUE THREAD

**Issue by alalazo** [LLNL/spack#36339]
### Steps to reproduce

Using this `compilers.yaml`:
```yaml
compilers:
- compiler:
    spec: gcc@11.0.0
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    flags: {}
    operating_system: ubuntu22.10
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
- compiler:
    spec: gcc@12.2.0
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    flags: {}
    operating_system: ubuntu22.10
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
```

### Error message

```console
$ spack spec  zlib%gcc@:11
Input spec
--------------------------------
zlib%gcc@:11

Concretized
--------------------------------
zlib@1.2.13%gcc@12.2.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-icelake
```

### Information on your system

* **Spack:** 0.20.0.dev0 (3d597e29bee1429d2b8d204016df40b575900d66)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
Should be solved by:
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index ca6d4baf13..df740445b4 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -915,6 +915,12 @@ error(2, "No valid version for '{0}' compiler '{1}' satisfies '@{2}'", Package,
      attr("node_compiler_version_satisfies", Package, Compiler, Constraint),
                 not compiler_version_satisfies(Compiler, Constraint, _).
 
+error(2, "No valid version for '{0}' compiler '{1}' satisfies '@{2}'", Package, Compiler, Constraint)
+  :- attr("node", Package),
+     attr("node_compiler_version_satisfies", Package, Compiler, Constraint),
+     not compiler_version_satisfies(Compiler, Constraint, ID),
+     node_compiler(Package, ID).
+
 % If the node is associated with a compiler and the compiler satisfy a constraint, then
 % the compiler associated with the node satisfy the same constraint
 attr("node_compiler_version_satisfies", Package, Compiler, Constraint)
```

**Comment by alalazo** [comment]
Another issue not caught by CI is with custom compiler versions (e.g. `gcc@foo`), which currently fails due to a type mismatch introduced in #35989 and not caught by unit tests:
```console
$ spack spec zlib %gcc@foo
==> Error: Can only build CompilerSpec from string or CompilerSpec. Found <class 'spack.compilers.gcc.Gcc'>
```
This is also solved in #36342
